File: src/app/model/Workout.ts
export enum DurationType {
    TIME = 'TIME',
    DISTANCE = 'DISTANCE',
    OPEN = 'OPEN'
}

export enum TargetType {
    HEART_RATE = 'HEART_RATE',
    POWER = 'POWER',
    SPEED = 'SPEED',
    CADENCE = 'CADENCE',
    OPEN = 'OPEN'
}


export class WorkoutStep {
    constructor(
        public durationType: DurationType,
        public durationValue: number,
        public targetType: TargetType,
        public targetLow: number | null = null,
        public targetHigh: number | null = null,
        public name: string,
    ) { }

    displayValue(value: number | null): string {
        if (this.targetType === TargetType.OPEN) {
            return 'Open';
        }
        if (this.targetType === TargetType.POWER) {
            return `${value}W`;
        }
        if (this.targetType === TargetType.HEART_RATE) {
            return `${value}bpm`;
        }
        if (this.targetType === TargetType.SPEED) {
            return `${value}m/s`;
        }
        if (this.targetType === TargetType.CADENCE) {
            return `${value}rpm`;
        }
        return 'N/A';
    }


    displayTargetLow(): string {
        return this.displayValue(this.targetLow);
    }
    displayTargetHigh(): string {
        return this.displayValue(this.targetHigh);
    }
    displayDuration(): string {
        if (this.durationType === DurationType.TIME) {
            const minutes = Math.floor(this.durationValue / 60);
            const seconds = Math.floor((this.durationValue % 60));
            return `${minutes}:${seconds < 10 ? '0' : ''}${seconds}`;
        } else if (this.durationType === DurationType.DISTANCE) {
            const kilometers = this.durationValue;
            return `${kilometers} km`;
        } else if (this.durationType === DurationType.OPEN) {
            return 'Open';
        } else {
            throw new Error(`Unknown duration type: ${this.durationType}`);
        }
    }

    /**
     * 
     * @param step e.g.
     * {
     * "durationType": "distance",
     * "durationValue": 500000,
     * "targetType": "power3s",
     * "customTargetValueLow": 1173,
     * "customTargetValueHigh": 1269,
     * "targetValue": 0,
     * "wktStepName": "Text to display",
     * "intensity": "interval",
     * "messageIndex": 0,
     * "durationDistance": 5000
     * }
     */
    static fromGarminFitFile(step: any): WorkoutStep {
        const durationType = WorkoutStep.parseDurationType(step.durationType);
        const targetType = WorkoutStep.parseTargetType(step.targetType);

        return new WorkoutStep(
            durationType,
            step.durationValue / 1000 || 0,
            targetType,
            this.parseCustomTargetValue(targetType, step.customTargetValueLow),
            this.parseCustomTargetValue(targetType, step.customTargetValueHigh),
            step.wktStepName || ''
        );

    }

    static parseDurationType(durationType: string): DurationType {
        switch (durationType.toLowerCase()) {
            case 'time':
                return DurationType.TIME;
            case 'distance':
                return DurationType.DISTANCE;
            case 'open':
                return DurationType.OPEN;
            default:
                throw new Error(`Unknown duration type: ${durationType}`);
        }
    }
    static parseTargetType(targetType: string): TargetType {
        if (targetType.toLowerCase().startsWith('power')) {
            return TargetType.POWER;
        } else if (targetType.toLowerCase().startsWith('heart')) {
            return TargetType.HEART_RATE;
        } else if (targetType.toLowerCase().startsWith('speed')) {
            return TargetType.SPEED;
        } else if (targetType.toLowerCase().startsWith('cadence')) {
            return TargetType.CADENCE;
        }
        return TargetType.OPEN;
    }

    static parseCustomTargetValue(targetType: TargetType, value: number): number | null {
        if (!value || value <= 0) {
            return null;
        }
        switch (targetType) {
            case TargetType.POWER:
                if (value < 1000) {
                    throw new Error("Relative power target not supported");
                }
                return value - 1000;
            case TargetType.HEART_RATE:
                if (value < 100) {
                    throw new Error("Relative heart rate target not supported");
                }
                return value - 100; // Convert to bpm
            case TargetType.SPEED:
                return value; // Speed is already in m/s
            case TargetType.CADENCE:
                return value; // Cadence is already in rpm
            default:
                return 0; // For OPEN or unknown types, return 0
        }
    }
}

export class Workout {
    constructor(
        public sport: string,
        public name: string,
        public steps: WorkoutStep[] = []
    ) { }

    static fromGarminFitFile(workout: any): Workout {
        const steps = workout.workoutStepMesgs.map((step: any) => WorkoutStep.fromGarminFitFile(step));
        return new Workout(
            workout.workoutMesgs[0]?.sport || 'cycling',
            workout.workoutMesgs[0]?.wktName || 'Unnamed Workout',
            steps
        );
    }
}

File: android/app/src/main/java/com/openbikecomputer/model/Workout.java
package com.openbikecomputer.model;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/**
 * A structured workout and its steps. Ported from {@code Workout.ts}.
 */
public class Workout {

    public enum DurationType {
        TIME, DISTANCE, OPEN
    }

    public enum TargetType {
        HEART_RATE, POWER, SPEED, CADENCE, OPEN
    }

    public static class WorkoutStep {
        public final DurationType durationType;
        /** Seconds (TIME) or kilometres (DISTANCE) — matching the original app. */
        public final double durationValue;
        public final TargetType targetType;
        public final Double targetLow;
        public final Double targetHigh;
        public final String name;

        public WorkoutStep(DurationType durationType, double durationValue, TargetType targetType,
                           Double targetLow, Double targetHigh, String name) {
            this.durationType = durationType;
            this.durationValue = durationValue;
            this.targetType = targetType;
            this.targetLow = targetLow;
            this.targetHigh = targetHigh;
            this.name = name;
        }

        public String displayValue(Double value) {
            switch (targetType) {
                case OPEN:
                    return "Open";
                case POWER:
                    return value + "W";
                case HEART_RATE:
                    return value + "bpm";
                case SPEED:
                    return value + "m/s";
                case CADENCE:
                    return value + "rpm";
                default:
                    return "N/A";
            }
        }

        public String displayTargetLow() {
            return displayValue(targetLow);
        }

        public String displayTargetHigh() {
            return displayValue(targetHigh);
        }

        public String displayDuration() {
            switch (durationType) {
                case TIME: {
                    int minutes = (int) Math.floor(durationValue / 60);
                    int seconds = (int) Math.floor(durationValue % 60);
                    return minutes + ":" + (seconds < 10 ? "0" : "") + seconds;
                }
                case DISTANCE:
                    return formatNumber(durationValue) + " km";
                case OPEN:
                    return "Open";
                default:
                    throw new IllegalStateException("Unknown duration type: " + durationType);
            }
        }

        /**
         * Build a step from the raw (FIT) field values, applying the same
         * relative-target resolution logic as the Angular app.
         */
        public static WorkoutStep fromRaw(String durationTypeRaw, double durationValueRaw,
                                          String targetTypeRaw,
                                          double customTargetValueLow, double customTargetValueHigh,
                                          String stepName,
                                          double powerThreshold, double heartRateThreshold) {
            DurationType durationType = parseDurationType(durationTypeRaw);
            TargetType targetType = parseTargetType(targetTypeRaw);

            Double threshold = targetType == TargetType.POWER ? powerThreshold
                    : targetType == TargetType.HEART_RATE ? heartRateThreshold : null;

            return new WorkoutStep(
                    durationType,
                    durationValueRaw / 1000.0,
                    targetType,
                    parseCustomTargetValue(targetType, customTargetValueLow, threshold),
                    parseCustomTargetValue(targetType, customTargetValueHigh, threshold),
                    stepName == null ? "" : stepName);
        }

        public static DurationType parseDurationType(String durationType) {
            if (durationType == null) {
                return DurationType.OPEN;
            }
            switch (durationType.toLowerCase(Locale.US)) {
                case "time":
                    return DurationType.TIME;
                case "distance":
                    return DurationType.DISTANCE;
                case "open":
                    return DurationType.OPEN;
                default:
                    // Unsupported duration types behave as open-ended (press lap).
                    return DurationType.OPEN;
            }
        }

        public static TargetType parseTargetType(String targetType) {
            if (targetType == null) {
                return TargetType.OPEN;
            }
            String t = targetType.toLowerCase(Locale.US);
            if (t.startsWith("power")) {
                return TargetType.POWER;
            } else if (t.startsWith("heart")) {
                return TargetType.HEART_RATE;
            } else if (t.startsWith("speed")) {
                return TargetType.SPEED;
            } else if (t.startsWith("cadence")) {
                return TargetType.CADENCE;
            }
            return TargetType.OPEN;
        }

        public static Double parseCustomTargetValue(TargetType targetType, double value, Double threshold) {
            if (value <= 0) {
                return null;
            }
            switch (targetType) {
                case POWER:
                    if (value < 1000) {
                        if (threshold == null || threshold == 0.0) {
                            throw new IllegalStateException("Relative power target not supported");
                        }
                        return threshold * value / 100.0;
                    }
                    return value - 1000;
                case HEART_RATE:
                    if (value < 100) {
                        if (threshold == null || threshold == 0.0) {
                            throw new IllegalStateException("Relative heart rate target not supported");
                        }
                        return threshold * value / 100.0;
                    }
                    return value - 100;
                case SPEED:
                    return value;
                case CADENCE:
                    return value;
                default:
                    return 0.0;
            }
        }

        private static String formatNumber(double v) {
            if (v == Math.floor(v)) {
                return Long.toString((long) v);
            }
            return Double.toString(v);
        }
    }

    public final String sport;
    public final String name;
    public final List<WorkoutStep> steps;

    public Workout(String sport, String name, List<WorkoutStep> steps) {
        this.sport = sport;
        this.name = name;
        this.steps = steps != null ? steps : new ArrayList<>();
    }
}
